hazelcast/sql.py

```diff
@@ -0,0 +1,1604 @@
+            # some unlucky timing, we might call _handle_fetch_response
+            # before returning, which could set self._fetch_future to
+            # None.
+            return future
+
+    def _handle_fetch_response(self, future):
+        """Handles the result of the fetch request, by either:
+
+        - setting it to exception, so that the future calls to
+          fetch fails immediately.
+        - setting it to next page, and setting self._fetch_future
+          to None so that the next fetch request might actually
+          try to fetch something from the server.
+
+        Args:
+            future (Future): The response from the server for
+            the fetch request.
+        """
+        try:
+            response = future.result()
+
+            response_error = self._handle_response_error(response["error"])
+            if response_error:
+                # There is a server side error sent to client.
+                self._on_fetch_error(response_error)
+                return
+
+            # The result contains the next page, as expected.
+            self._on_fetch_response(response["row_page"])
+        except Exception as e:
+            # Something went bad, we couldn't get response from
+            # the server, invocation failed.
+            self._on_fetch_error(self._sql_service.re_raise(e, self._connection))
+
+    def _on_fetch_error(self, error):
+        """Sets the fetch future with exception, but not resetting it
+        so that the next fetch request fails immediately.
+
+        Args:
+            error (Exception): The error.
+        """
+        with self._lock:
+            self._fetch_future.set_exception(error)
+
+    def _on_fetch_response(self, page):
+        """Sets the fetch future with the next page,
+        resets it, and if this is the last page,
+        marks the result as closed.
+
+        Args:
+            page (_SqlPage): The next page.
+        """
+        with self._lock:
+            self._fetch_future.set_result(page)
+            if page.is_last:
+                # This is the last page, there is nothing
+                # more on the server.
+                self._mark_closed()
+
+            self._fetch_future = None
+
+    def _handle_execute_response(self, future):
+        """Handles the result of the execute request, by either:
+
+        - setting it to an exception so that the dependent methods
+          (iterator, update_count etc.) fails immediately
+        - setting it to an execute response
+
+        Args:
+            future (Future):
+        """
+        try:
+            response = future.result()
+
+            response_error = self._handle_response_error(response["error"])
+            if response_error:
+                # There is a server-side error sent to the client.
+                self._on_execute_error(response_error)
+                return
+
+            row_metadata = response["row_metadata"]
+            if row_metadata is not None:
```

Comment:
Just a note: not setting `row_metadata` key instead of setting it to `None` may be better.

Reply:
if I understand your note correctly, you mean we shouldn't set `response["row_metadata"]` to `None`. Instead, there shouldn't be such a key in response dict. 

That requires a codec template change, and personally, I would like to leave this as it is to distinguish `None` value with no response parameter named as such cases.

If I understand your note incorrectly, please comment